### DISCUSSION TITLE
Problem in Page Binding of "Right to Left"

### DISCUSSION THREAD

**Discussion by mlove4u** [pypdfium2-team/pypdfium2#112] [General]
https://github.com/pypdfium2-team/pypdfium2/blob/8ffcafc5fe8db8de6ed9f6e6279bf2e8933b8916/src/pypdfium2/_helpers/text_extractor.py#L83

When pdf page binding was set to "right to left", the text will be reversed after extracted.
I [reported](https://bugs.chromium.org/p/pdfium/issues/detail?id=1728&q=left%20to%20right&can=2) this problem to the pdfium team last year, but it doesn't seem to be solved so far. Although I don't know c++, I guess this can't be solved only inside pypdfium, right?
Test code:
```python3
import pypdfium2 as pdfium


def mytest(filename):
    doc = pdfium.PdfDocument(filename)
    page = doc.get_page(0)
    textpage = pdfium.PdfTextPage(page)
    text = textpage.get_text(45, 515, 375, 550) # pink rectangle
    return text


file1 = "left_to_right.pdf"
file2 = "right_to_left.pdf"
print(mytest(file1))  # 'To Every Thing There Is A Season'
# words are reversed
print(mytest(file2))  # 'Season A Is There Thing Every To'
```
[left_to_right.pdf](https://github.com/pypdfium2-team/pypdfium2/files/8819762/left_to_right.pdf)
[right_to_left.pdf](https://github.com/pypdfium2-team/pypdfium2/files/8819764/right_to_left.pdf)

**Comment by mara004** [comment]
I suppose you are referring to `ViewerPreferences.Direction` (`L2R` or `R2L`) as outlined in the [PDF 1.7 Reference](https://opensource.adobe.com/dc-acrobat-sdk-docs/standards/pdfstandards/pdf/PDF32000_2008.pdf) on page 371.
Perhaps you could use [`FPDF_VIEWERREF_GetName()`](https://pdfium.googlesource.com/pdfium/+/15d326e3d3fbbf178fde4c22ec8860f08b54362c/public/fpdfview.h#1231) to obtain the reading order and reverse the words if it's `R2L`?
That said, it would of course be better if the issue could be fixed in PDFium.

**Reply by mara004** [reply]
I've tried to create a patch for PDFium that would just remove the inversion: [0001-Remove-R2L-Inversion.patch.txt](https://github.com/pypdfium2-team/pypdfium2/files/8825313/0001-Remove-R2L-Inversion.patch.txt)
Here's the output with the patch: [output.txt](https://github.com/pypdfium2-team/pypdfium2/files/8825315/output.txt)
However, the change causes at least one failing test case, namely `FPDFTextEmbedderTest.TextHebrewMirrored`.
Apparently, the inversion is necessary in some cases to compensate flip, so disabling it completely as I did can't be the solution...

**Reply by mlove4u** [reply]
@mara004  Awesome. 
Yes, it is "ViewerPreferences.Direction". I often export pdf from InDesign document, and it is called "page binding" in InDesign.
I used to use pymupdf and xpdf to extract text from pdf documents(these two works fine in R2L pdf), but because of their license, and then switched to use pdfium, and happened to stumble upon this problem.
Thanks.

vertical text layout sample pdf:
[vertical_text_layout.pdf](https://github.com/pypdfium2-team/pypdfium2/files/8828966/vertical_text_layout.pdf)
